File: examples/code_restart/native_code/code_restart.f90
module mod_code_restart

integer :: code_state = 0

contains

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_restart(equilibrium_in, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_is_valid

        implicit none

        type(ids_equilibrium):: equilibrium_in
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message
        integer :: i

        ! INITIAL DISPLAY
        write(*,*) ' '
        write(*,*) '======================================='
        write(*,*) 'START OF PHYSICS CODE'

        ! INITIALISATION OF ERROR FLAG
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'code_restart: OK'

        write(*,*) 'Starting from: ', code_state

        do i = 1, 20
            ! COMPUTATIONS
            code_state = code_state + 1
        end do

        write(*,*) 'Counting to: ', code_state

        ! FINAL DISPLAY
        write(*,*) 'END OF PHYSICS CODE'
        write(*,*) '======================================='
        write(*,*) ' '

        end subroutine code_restart

    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine get_code_state (state_str, status_code, status_message)

        implicit none
        character(len=:), allocatable, intent(out) :: state_str
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        allocate(character(50):: state_str)
        write(state_str,*) code_state



        write(*,*) '======================================='
        write(*,*) 'Code lifecycle: GET CODE STATE called'
        write(*,*) 'STATE is :', state_str
        write(*,*) '======================================='

    end subroutine get_code_state

    subroutine restore_code_state (state_str, status_code, status_message)
    !
    !    INITIALISATION SUBROUTINE
    !
        implicit none
        character(len=:), allocatable, intent(in) :: state_str
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        read(state_str , *) code_state
        write(*,*) '======================================='
        write(*,*) 'Code lifecycle: RESTORE STATE called'
        write(*,*) 'STATE TO BE RESTORED :', code_state
        write(*,*) '======================================='

    end subroutine restore_code_state

end module mod_code_restart

File: examples/code_restart_cpp/native_code/code_restart.cpp
#if AL_MAJOR == 5
    #include "ALClasses.h"
#elif AL_MAJOR == 4
    #include "UALClasses.h"
#else
    #warning Could not find AL_MAJOR variable. Assuming AL version = 5.x.x.
    #include "ALClasses.h"
#endif

#include "code_restart.h"

// =======================================
//             GET STATE
//=======================================

int code_state = 0;

void get_code_state( std::string& state_out, int& status_code, std::string& status_message)
{
    status_code = 0;

    status_message = "INITIALISATION: OK";
    state_out = std::to_string(code_state);

    std::cout << "=======================================================" << std::endl;
    std::cout << "Code restart CPP: GET STATE called" << std::endl;
    std::cout << "STATE is : " << state_out << std::endl;
    std::cout << "=======================================================" << std::endl;
}

// =======================================
//             SET STATE
//=======================================
void restore_code_state( std::string state, int& status_code, std::string& status_message)
{
    status_code = 0;
    status_message = "FINALISATION: OK";

    code_state = std::stoi( state );
    std::cout << "=======================================================" << std::endl;
    std::cout << "Code lifecycle CPP: RESTORE STATE called" << std::endl;
    std::cout << "STATE TO BE RESTORED : " << code_state << std::endl;
    std::cout << "=======================================================" << std::endl;
}

// =======================================
//             MAIN
//=======================================
void code_restart(const IdsNs::IDS::equilibrium& in_equilibrium,
                     int& status_code, std::string& status_message)
{

    // INITIALISATION OF ERROR FLAG
    status_code = 0;
    
    // INITIAL DISPLAY
    std::cout <<  "=======================================" << std::endl;
    std::cout <<  "START OF PHYSICS CODE" << std::endl;
    

    std::cout <<  "Starting from: " << code_state << std::endl;



    for (int i = 0; i < 20; i++)
    {
        // COMPUTATIONS
        code_state++;
    }


    std::cout <<  "Counting to : " << code_state << std::endl;

    // INITIALISATION OF STATUS INFO
    status_message = "Status info of code_restart CPP";

    
    // FINAL DISPLAY
    std::cout <<  "END OF PHYSICS CODE" << std::endl;
    std::cout <<  "=======================================" << std::endl;
    std::cout <<  " " << std::endl;
}
